src/network/http.cpp

```diff
@@ -31,198 +31,299 @@ void http_dispose() { }
+        free(response->body);
 
-	curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, request->method);
-	curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
-	curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, true);
-	curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, true);
-	curl_easy_setopt(curl, CURLOPT_USERAGENT, OPENRCT2_USER_AGENT);
-	curl_easy_setopt(curl, CURLOPT_URL, request->url);
-	curl_easy_setopt(curl, CURLOPT_WRITEDATA, &writeBuffer);
-	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, http_request_write_func);
+    free(response);
+}
 
-	curlResult = curl_easy_perform(curl);
+const char *http_get_extension_from_url(const char *url)
+{
+    const char *extension = strrchr(url, '.');
+
+    // Assume a save file by default if no valid extension can be determined
+    if (extension == NULL || strchr(extension, '/') != NULL) {
+        return ".sv6";
+    } else {
+        return extension;
+    }
+}
 
-	if (request->body != NULL) {
-		free(readBuffer.ptr);
+bool http_download_park(const char *url, char tmpPath[L_tmpnam])
+{
+	// Download park to buffer in memory
+	http_request_t request;
+	request.url = url;
+	request.method = "GET";
+	request.type = HTTP_DATA_NONE;
+
+	http_response_t *response = http_request(&request);
+
+	if (response == NULL || response->status_code != 200) {
+		fprintf(stderr, "Failed to download '%s'", request.url);
+		if (response != NULL) {
+			http_request_dispose(response);
+		}
+		return false;
 	}
 
-	if (curlResult != CURLE_OK) {
-		log_error("HTTP request failed: %s.", curl_easy_strerror(curlResult));
-		if (writeBuffer.ptr != NULL)
-			free(writeBuffer.ptr);
-
-		return NULL;
+	// Generate temporary filename that includes the original extension
+	if (tmpnam(tmpPath) == NULL) {
+		fprintf(stderr, "Failed to generate temporary filename for downloaded park '%s'", request.url);
+		http_request_dispose(response);
+		return false;
 	}
+	size_t remainingBytes = sizeof(tmpPath) - strlen(tmpPath);
 
-	long httpStatusCode;
-	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpStatusCode);
-
-	curl_easy_cleanup(curl);
+	const char *ext = http_get_extension_from_url(request.url);
+	strncat(tmpPath, ext, remainingBytes);
 
-	// Null terminate the response buffer
-	writeBuffer.length++;
-	writeBuffer.ptr = (char*)realloc(writeBuffer.ptr, writeBuffer.length);
-	writeBuffer.capacity = writeBuffer.length;
-	writeBuffer.ptr[writeBuffer.length - 1] = 0;
+	// Store park in temporary file and load it (discard ending NUL in response body)
+	FILE* tmpFile = fopen(tmpPath, "wb");
 
-	response = NULL;
-
-	// Parse as JSON
-	json_t *root;
-	json_error_t error;
-	root = json_loads(writeBuffer.ptr, 0, &error);
-	if (root != NULL) {
-		response = (http_json_response*)malloc(sizeof(http_json_response));
-		response->tag = request->tag;
-		response->status_code = (int)httpStatusCode;
-		response->root = root;
+	if (tmpFile == NULL) {
+		fprintf(stderr, "Failed to write downloaded park '%s' to temporary file", request.url);
```

Comment:
Use `Console::Error::WriteLine` which will also add the new line character for you which you have forgotten to add.